src/components/Identity/createMigration.js

```diff
@@ -10,15 +10,65 @@ OF ANY KIND, either express or implied. See the License for the specific languag
 governing permissions and limitations under the License.
 */
 
-import { getApexDomain, cookieJar } from "../../utils";
+import { getApexDomain, cookieJar, isObject, isFunction } from "../../utils";
 // TODO: We are already retrieving the apex in core; find a way to reuse it.
 // Maybe default the domain in the cookieJar to apex while allowing overrides.
 const apexDomain = getApexDomain(window, cookieJar);
 
-export default ({ orgId, consent }) => {
+export default ({ orgId, consent, logger }) => {
   const amcvCookieName = `AMCV_${orgId}`;
+  const Visitor = window.Visitor;
+  const doesVisitorExist =
+    isFunction(Visitor) && isFunction(Visitor.getInstance);
+
+  const awaitVisitorOptIn = () => {
+    return new Promise(resolve => {
+      if (isObject(window.adobe) && isObject(window.adobe.optIn)) {
+        const optInOld = window.adobe.optIn;
+        logger.log(
+          "Delaying request while waiting for legacy opt in to let Visitor retrieve ECID from server."
+        );
+        optInOld.fetchPermissions(() => {
+          if (optInOld.isApproved([optInOld.Categories.ECID])) {
+            logger.log(
+              "Received legacy opt in approval to let Visitor retrieve ECID from server."
+            );
+
+            resolve();
+          }
+        }, true);
+      } else {
+        resolve();
+      }
+    });
+  };
+
+  const getVisitorECID = () => {
+    return awaitVisitorOptIn().then(() => {
+      return new Promise(resolve => {
+        const visitor = Visitor.getInstance(orgId, {});
+        visitor.getMarketingCloudVisitorID(resolve, true);
+      });
+    });
+  };
 
   return {
+    getEcidFromLegacy() {
+      const ecid = this.getEcidFromLegacyCookies();
+
+      if (ecid) {
+        return new Promise(resolve => {
+          resolve(ecid);
+        });
+      }
+
+      if (doesVisitorExist) {
+        return getVisitorECID();
+      }
+      return new Promise(resolve => {
```

Comment:
You can return something falsey here (false, null, or undefined) and then the addEcidToPayload won't run.